packages/lsp-tools/src/doenet-to-markdown/index.ts

```diff
@@ -0,0 +1,254 @@
+import { DastNodes, DastNodesV6, DastRoot, toXml } from "@doenet/parser";
+import type {
+    Root as MdastRoot,
+    Nodes as MdastNodes,
+    RootContent as MdastRootContent,
+    PhrasingContent as MdastPhrasingContent,
+    BlockContent as MdastBlockContent,
+    ListItem as MdastListItem,
+} from "mdast";
+import { toMarkdown } from "mdast-util-to-markdown";
+import { DoenetSourceObject } from "../doenet-source-object";
+import { condenseBlankPars, trimLeadingWhitespace } from "./trim";
+import { sum } from "./utils";
+
+const DIVISIONS = [
+    "part",
+    "chapter",
+    "section",
+    "subsection",
+    "subsubsection",
+    "paragraphs",
+    // Not officially divisions, but we treat them as divisions
+    // to get more reliable headers
+    "example",
+    "aside",
+];
+
+export function doenetToMarkdown(source: DoenetSourceObject | string) {
+    const sourceObj =
+        typeof source === "string" ? new DoenetSourceObject(source) : source;
+    const dast = sourceObj.dast as DastRoot;
+    let mdast = condenseBlankPars(doenetToMdast(dast, sourceObj));
+
+    // Make sure there are paragraphs between consecutive headers.
+    // Otherwise blank lines will not be rendered between consecutive headers.
+    mdast = mdast.flatMap((node, i) => {
+        const nextNode = mdast[i + 1];
+        if (node.type === "heading" && nextNode?.type === "heading") {
+            return [
+                node,
+                {
+                    type: "paragraph",
+                    children: [{ type: "text", value: "\n\n" }],
+                },
+            ];
+        }
+        if (node.type === "heading" && nextNode?.type !== "paragraph") {
+            return [
+                node,
+                {
+                    type: "paragraph",
+                    children: [{ type: "text", value: "\n\n" }],
+                },
+            ];
+        }
+        return [node];
+    });
+
+    let text = toMarkdown({ type: "root", children: mdast });
+    // Occurrences of `&#x20;` at the start of a line are probably a mistake, so trim them.
+    text = text.replace(/^&#x20;/gm, "");
```

Comment:
Where are the `&#x20;` coming from? Do we care that 
```
<section>
    <title>title 1</title>
    hello 1
    <section>
        <title>title 2</title>
        hello 2
    </section>
</section>
```
leads to stray values of `&#x20;` lying around. They go away if you add `<p>`s around the text.

Reply:
I vote for ignoring them for now. They are there because they are the text encoding of newlines. Since we don't know where the `<p>`s start and stop, I decided not to trim whitespace from the start/end of nodes, lest everything gets shoved to one line. A consequence is sometimes a `\n` shows up in a `type: "text"` node which gets turned to `&#x20;`